guardian/typerighter
# Sort by rule order

## What does this change?



This adds a `rule_order` column to the `rules_draft` table, which will be used to determine prioritisation of rules that may conflict.

Previously (before the TR manager), rule order was determined by the order of the rule in the spreadsheet. This order has been maintained in the database by the rule id - `rule_order` in the `rules_live` table was determined by copying this id when a rule is published.

When there is already data in the database, the logic added in this PR copies the `id` in `rules_draft` into a new `rule_order` column (done as an evolution). When a new rule is added, a query will find the latest rule order and increment this by 1. If there is no data in the database, the `rule_order` will be assigned according to the spreadsheet row index, thus maintaining the spreadsheet order.

Trello card is [here](https://trello.com/c/qhEI4OVe/1534-sort-by-rule-order-in-table).

### Next
Sort by rule order on the client. In future we'll look at implementing a way for users to rearrange the order of conflicting rules.

## How to test



This can be tested locally by running `sbt`, and running the tests with:
```
project ruleManager
test
```
 All tests should pass and the latest evolution should be applied. If your database already contained data, all rows in the `rules_draft` column should be populated, with the values of `rules_draft` matching the values of the rule's `id`.

To test the logic when the database starts without any tables, drop your local database tables, run the manager app and press the big red `Destroy all rules in the manager...` button in the UI. The `rules_draft` table should be created with a `rule_order` column. The value of `rule_order` won't necessarily match the `id` as the `id` auto increment counter does not reset automatically.

This has been tested on CODE. Instructions [here](https://github.com/guardian/typerighter/blob/main/README.md#developer-how-tos) can be followed to connect to the CODE database.
